File: ios/Classes/WiseaiSdkPlugin.swift
import Flutter
import UIKit
import WiseAISDK

/**
 * WiseAI Flutter Bridge
 *
 * Design contract (do not change without updating the Flutter side):
 *
 * The native bridge forwards SDK responses verbatim. It never parses the
 * SDK's `status` / `code` / `message` / `sessionId`, never reshapes the
 * JSON, and never decides what is "success" vs "error" — Flutter does that.
 *
 * The bridge always sends one of three event sources, plus an optional
 * event-level sessionId:
 *   { "source": "sdk",         "rawData": <verbatim SDK JSON, decrypted if needed>, "sessionId"?: "..." }
 *   { "source": "cancelled",   "rawData": "",                                       "sessionId"?: "..." }
 *   { "source": "bridgeError", "rawData": <human message — only when SDK never ran>, "sessionId"?: "..." }
 *
 * "bridgeError" means the SDK was never reached. SDK-reported errors
 * (session expired, OCR mismatch, etc.) come back via "sdk" because the
 * SDK itself produced them.
 */

/// `WiseaiSdkPlugin` is `@objc`-exported (it's a `FlutterPlugin`), which means
/// Xcode generates an Objective-C compatibility header for it that Runner
/// imports across the pod boundary. If that public class conformed directly
/// to `WiseAiDelegate`/`FaceVerifyDelegate`, the generated header would need
/// to reference those WiseAISDK protocol names — but WiseAISDK isn't
/// guaranteed to be visible at the point Runner consumes that header
/// (unlike Flutter.framework, which every pod can assume is imported),
/// causing "Cannot find protocol declaration for 'FaceVerifyDelegate'".
/// Keeping the SDK-protocol conformance on this private, non-exported
/// object instead avoids that entirely.
private class WiseAiDelegateHandler: NSObject, WiseAiDelegate, FaceVerifyDelegate {
  weak var plugin: WiseaiSdkPlugin?

  init(plugin: WiseaiSdkPlugin) {
    self.plugin = plugin
    super.init()
  }

  func getSessionIdAndEncryptionConfig(_ sessionIdAndEncryptionConfig: String) {
    plugin?.handleSessionIdAndEncryptionConfig(sessionIdAndEncryptionConfig)
  }

  func onEkycComplete(_ jsonResult: String)  { plugin?.handleEkycComplete(jsonResult) }
  func onEkycException(_ jsonResult: String) { plugin?.handleEkycException(jsonResult) }
  func onEkycCancelled()                     { plugin?.handleEkycCancelled() }

  func getFaceVerifySessionIdAndEncryptionConfig(_ sessionIdAndEncryptionConfig: String) {
    plugin?.handleFaceVerifySessionIdAndEncryptionConfig(sessionIdAndEncryptionConfig)
  }

  func onFaceVerifyComplete(_ jsonResult: String)  { plugin?.handleFaceVerifyComplete(jsonResult) }
  func onFaceVerifyException(_ jsonResult: String) { plugin?.handleFaceVerifyException(jsonResult) }
  func onFaceVerifyCancelled()                     { plugin?.handleFaceVerifyCancelled() }
}

@objc(WiseaiSdkPlugin)
public class WiseaiSdkPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

  private var wiseAiApp: WiseAiApp?
  private var encryptionConfig: [String: Any] = [:]
  private var eventSink: FlutterEventSink?
  private var delegateHandler: WiseAiDelegateHandler?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = WiseaiSdkPlugin()
    instance.delegateHandler = WiseAiDelegateHandler(plugin: instance)

    let methodChannel = FlutterMethodChannel(name: "WiseAiMethods",
                                              binaryMessenger: registrar.messenger())
    methodChannel.setMethodCallHandler(instance.handle)

    let eventChannel = FlutterEventChannel(name: "com.wiseai.wiseai_sdk_plugin/events",
                                            binaryMessenger: registrar.messenger())
    eventChannel.setStreamHandler(instance)
  }

  // MARK: - MethodChannel

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "performMykadEkyc":
      performMykadEkyc(args: args)
      result(nil)
    case "performPassportNFCEkyc":
      performPassportNFCEkyc(args: args)
      result(nil)
    case "performFaceVerify":
      performFaceVerify(args: args)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - SDK Operations

  private func performMykadEkyc(args: [String: Any]) {
    let apiToken = args["apiToken"] as? String ?? ""
    let apiURL = args["apiURL"] as? String ?? ""
    let language = args["language"] as? String ?? "EN"
    let isEncrypt = args["isEncrypt"] as? Bool ?? false

    wiseAiApp = WiseAiApp(ekycApiToken: apiToken, ekycApiURL: apiURL)
    // or initialize SDK with extraParam
    // let extraParam = buildExtraParamJson(args["extraParam"] as? [String: String])
    // wiseAiApp = WiseAiApp(ekycApiToken: apiToken, ekycApiURL: apiURL, extraParam: extraParam)
    wiseAiApp?.delegate = delegateHandler
    wiseAiApp?.setLanguage(language)
    wiseAiApp?.performEkyc(isEncrypt: isEncrypt)
  }

  private func performPassportNFCEkyc(args: [String: Any]) {
    let apiToken = args["apiToken"] as? String ?? ""
    let apiURL = args["apiURL"] as? String ?? ""
    let language = args["language"] as? String ?? "EN"
    let isNFC = args["isNFC"] as? Bool ?? false
    let isEncrypt = args["isEncrypt"] as? Bool ?? false

    wiseAiApp = WiseAiApp(ekycApiToken: apiToken, ekycApiURL: apiURL)
    // or initialize SDK with extraParam
    // let extraParam = buildExtraParamJson(args["extraParam"] as? [String: String])
    // wiseAiApp = WiseAiApp(ekycApiToken: apiToken, ekycApiURL: apiURL, extraParam: extraParam)
    wiseAiApp?.delegate = delegateHandler
    wiseAiApp?.setLanguage(language)
    wiseAiApp?.performPassportEkyc(isEncrypt: isEncrypt, isNFC: isNFC)
  }

  private func performFaceVerify(args: [String: Any]) {
    let apiToken = args["apiToken"] as? String ?? ""
    let apiURL = args["apiURL"] as? String ?? ""
    let language = args["language"] as? String ?? "EN"
    let isExportFace = args["isExportFace"] as? Bool ?? false
    let isActiveLiveness = args["isActiveLiveness"] as? Bool ?? false
    let isEncrypt = args["isEncrypt"] as? Bool ?? false

    guard let faceImageBase64 = args["faceImageBase64"] as? String else {
      sendBridgeError("Face Verify: faceImageBase64 is required")
      return
    }
    guard let rawBinaryImageData = Data(base64Encoded: faceImageBase64) else {
      sendBridgeError("Face Verify: failed to decode base64 image")
      return
    }

    wiseAiApp = WiseAiApp(ekycApiToken: apiToken, ekycApiURL: apiURL)
    wiseAiApp?.delegate = delegateHandler
    wiseAiApp?.faceVerifyDelegate = delegateHandler
    wiseAiApp?.setLanguage(language)
    wiseAiApp?.performImageFaceVerify(
      rawBinaryImageData: rawBinaryImageData,
      isExportFace: isExportFace,
      isActiveLiveness: isActiveLiveness,
      isEncrypt: isEncrypt
    )
  }

  // MARK: - Helpers

  private func buildExtraParamJson(_ map: [String: String]?) -> String {
    guard let map = map,
          let data = try? JSONSerialization.data(withJSONObject: map),
          let str = String(data: data, encoding: .utf8) else {
      return "{}"
    }
    return str
  }

  /// Decrypt if encrypted; on any failure, return the original string so the
  /// caller still has something with sessionId to forward. Never throws.
  private func decryptIfNeeded(_ result: String) -> String {
    guard let decrypted = wiseAiApp?.decryptResult(encryptedResult: result, encryptionConfig: encryptionConfig),
          !decrypted.isEmpty,
          decrypted != "Failed to decrypt data." else {
      return result
    }
    return decrypted
  }

  private func storeEncryptionConfig(from sessionData: Any) {
    if let configDict = sessionData as? [String: Any],
       let config = configDict["encryptionConfig"] as? [String: String] {
      encryptionConfig = config
      return
    }
    if let jsonString = sessionData as? String,
       let data = jsonString.data(using: .utf8),
       let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
       let config = json["encryptionConfig"] as? [String: String] {
      encryptionConfig = config
    }
  }

  // MARK: - Event Senders (the ONLY three exits to Flutter)

  private func sendSdkResult(_ rawData: String) { sendEvent(source: "sdk", rawData: rawData) }
  private func sendCancelled()                  { sendEvent(source: "cancelled", rawData: "") }
  private func sendBridgeError(_ message: String) {
    print("[WiseAiBridge] bridgeError: \(message)")
    sendEvent(source: "bridgeError", rawData: message)
  }

  private func sendEvent(source: String, rawData: String) {
    guard let sink = eventSink else {
      print("[WiseAiBridge] eventSink is nil — dropping event source=\(source)")
      return
    }
    sink(["source": source, "rawData": rawData])
  }

  // MARK: - Internal delegate callback handlers
  //
  // Called by WiseAiDelegateHandler (not exposed to Objective-C). Not
  // `private` — a sibling class in this file needs to call these.

  func handleSessionIdAndEncryptionConfig(_ sessionIdAndEncryptionConfig: String) {
    storeEncryptionConfig(from: sessionIdAndEncryptionConfig)
  }

  func handleEkycComplete(_ jsonResult: String)  { sendSdkResult(decryptIfNeeded(jsonResult)) }
  func handleEkycException(_ jsonResult: String) { sendSdkResult(jsonResult) }
  func handleEkycCancelled()                     { sendCancelled() }

  func handleFaceVerifySessionIdAndEncryptionConfig(_ sessionIdAndEncryptionConfig: String) {
    storeEncryptionConfig(from: sessionIdAndEncryptionConfig)
  }

  func handleFaceVerifyComplete(_ jsonResult: String)  { sendSdkResult(decryptIfNeeded(jsonResult)) }
  func handleFaceVerifyException(_ jsonResult: String) { sendSdkResult(jsonResult) }
  func handleFaceVerifyCancelled()                     { sendCancelled() }

  // MARK: - FlutterStreamHandler

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    self.eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    self.eventSink = nil
    return nil
  }
}

File: android/src/main/kotlin/com/example/wiseai_sdk_plugin/WiseaiSdkPlugin.kt
package com.example.wiseai_sdk_plugin

import android.app.Activity
import android.content.Intent
import android.os.Build
import android.util.Base64
import android.util.Log
import androidx.annotation.NonNull
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.wiseai.ekyc110.WiseAiApp
import com.wiseai.ekyc110.ekyc.Ekyc
import com.wiseai.ekyc110.ekyc.PassportNFCEkyc
import com.wiseai.ekyc110.ekyc.WiseAiFaceVerify
import com.wiseai.ekyc110.helper.SessionCallback
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * WiseAI Flutter Bridge
 *
 * Design contract (do not change without updating the Flutter side):
 *
 * The native bridge forwards SDK responses verbatim. It never parses the
 * SDK's `status` / `code` / `message` / `sessionId`, never reshapes the
 * JSON, and never decides what is "success" vs "error" — Flutter does that.
 *
 * The bridge always sends one of three event sources, plus an optional
 * event-level sessionId:
 *   { "source": "sdk",         "rawData": <verbatim SDK JSON, decrypted if needed>, "sessionId"?: "..." }
 *   { "source": "cancelled",   "rawData": "",                                       "sessionId"?: "..." }
 *   { "source": "bridgeError", "rawData": <verbatim message — no SDK eKYC payload>, "sessionId"?: "..." }
 *
 * The event-level "sessionId" is captured from WiseAiApp.startNewSession
 * (SDK v2.7.7+ returns it immediately upon session creation, regardless
 * of encryption). It lets Flutter correlate cancelled / bridgeError events
 * to a session even when no SDK eKYC payload is produced.
 *
 * "sdk" carries the SDK's structured eKYC response from getResult() — both
 * the success and the error variant.
 *
 * "bridgeError" covers every case where there is no structured SDK response
 * to forward — bridge-side failures (missing args, decode errors), SDK
 * callbacks that return only a String (SessionCallback.onError), or the SDK
 * activity exiting without a payload. The rawData carries whatever message
 * we have, verbatim. The bridge never invents JSON.
 */
class WiseaiSdkPlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler,
    ActivityAware,
    PluginRegistry.ActivityResultListener {

    private val METHOD_CHANNEL = "WiseAiMethods"
    private val EVENT_CHANNEL = "com.wiseai.wiseai_sdk_plugin/events"
    private val ACTIVITY_EKYC = 1001
    private val ACTIVITY_PASSPORT_NFC_EKYC = 1002
    private val ACTIVITY_FACE_VERIFY = 1003
    private val TAG = "WiseAiBridge"

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var activity: Activity? = null
    private var eventSink: EventChannel.EventSink? = null

    private val wiseAiApp: WiseAiApp = WiseAiApp()

    private lateinit var encryptionAlgorithm: String
    private lateinit var key: String
    private lateinit var iv: String
    private lateinit var padding: String
    private lateinit var mode: String
    private var ifEncryption: Boolean = false

    /**
     * SessionId returned by WiseAiApp.startNewSession (SDK v2.7.7+ returns it
     * regardless of encryption). Reset at the start of each flow; attached to
     * cancelled / bridgeError events so Flutter can correlate them to the
     * session even when no SDK eKYC payload is produced.
     */
    private var currentSessionId: String = ""

    // --- FlutterPlugin ---

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    // --- ActivityAware ---

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // --- EventChannel.StreamHandler ---

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.eventSink = events
        Log.d(TAG, "EventChannel: Flutter started listening")
    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
        Log.d(TAG, "EventChannel: Flutter stopped listening")
    }

    // --- MethodCallHandler ---

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getPlatformVersion" -> result.success("Android ${Build.VERSION.RELEASE}")
            "performMykadEkyc" -> {
                performMykadEkyc(call.arguments as? Map<String, Any> ?: emptyMap())
                result.success(null)
            }
            "performPassportNFCEkyc" -> {
                performPassportNFCEkyc(call.arguments as? Map<String, Any> ?: emptyMap())
                result.success(null)
            }
            "performFaceVerify" -> {
                performFaceVerify(call.arguments as? Map<String, Any> ?: emptyMap())
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // MARK: - SDK Operations

    private fun performMykadEkyc(args: Map<String, Any>) {
        val currentActivity = activity
        if (currentActivity == null) {
            sendBridgeError("Activity not available")
            return
        }
        currentSessionId = ""

        val apiToken = args["apiToken"] as? String ?: ""
        val apiURL = args["apiURL"] as? String ?: ""
        val language = args["language"] as? String ?: "EN"
        ifEncryption = args["isEncrypt"] as? Boolean ?: false
        val extraParamMap = args["extraParam"] as? Map<String, String>

        val extraParam = JsonObject()
        extraParamMap?.forEach { (k, v) -> extraParam.addProperty(k, v) }

        wiseAiApp.init(currentActivity, apiToken, apiURL)
        // or initialize SDK with extraParam
        //wiseAiApp.initWithExtraParam(currentActivity, apiToken, apiURL, extraParam)

        val intent = Intent(currentActivity, Ekyc::class.java).apply {
            putExtra("COUNTRY_CODE", "MYS")
            putExtra("ID_TYPE", "ID")
            putExtra("EXPORT_FACE", true)
            putExtra("CAMERA_FACING", "FRONT")
            putExtra("IF_ENCRYPTION", ifEncryption)
            putExtra("LANGUAGE_CODE", language)
        }

        WiseAiApp.startNewSession(ifEncryption, object : SessionCallback {
            override fun onComplete(data: Any?) {
                handleSessionCallback(data)
                currentActivity.startActivityForResult(intent, ACTIVITY_EKYC)
            }
            override fun onError(exception: String) {
                // Session never started → no structured SDK eKYC response exists.
                // Forward the SDK's message verbatim; bridge does not invent JSON.
                sendBridgeError(exception)
            }
        })
    }

    private fun performPassportNFCEkyc(args: Map<String, Any>) {
        val currentActivity = activity
        if (currentActivity == null) {
            sendBridgeError("Activity not available")
            return
        }
        currentSessionId = ""
        val apiToken = args["apiToken"] as? String ?: ""
        val apiURL = args["apiURL"] as? String ?: ""
        val language = args["language"] as? String ?: "EN"
        ifEncryption = args["isEncrypt"] as? Boolean ?: false
        val extraParamMap = args["extraParam"] as? Map<String, String>

        val extraParam = JsonObject()
        extraParamMap?.forEach { (k, v) -> extraParam.addProperty(k, v) }

        wiseAiApp.init(currentActivity, apiToken, apiURL)
        // or initialize SDK with extraParam
        // wiseAiApp.initWithExtraParam(currentActivity, apiToken, apiURL, extraParam)

        val intent = Intent(currentActivity, PassportNFCEkyc::class.java).apply {
            putExtra("COUNTRY_CODE", "")
            putExtra("ID_TYPE", "PASSPORT")
            putExtra("EXPORT_FACE", true)
            putExtra("TIMEOUT_PERIOD", 15)
            putExtra("CAMERA_FACING", "FRONT")
            putExtra("IF_ENCRYPTION", ifEncryption)
            putExtra("LANGUAGE_CODE", language)
        }

        WiseAiApp.startNewSession(ifEncryption, object : SessionCallback {
            override fun onComplete(data: Any?) {
                handleSessionCallback(data)
                currentActivity.startActivityForResult(intent, ACTIVITY_PASSPORT_NFC_EKYC)
            }
            override fun onError(exception: String) {
                sendBridgeError(exception)
            }
        })
    }

    private fun performFaceVerify(args: Map<String, Any>) {
        val currentActivity = activity
        if (currentActivity == null) {
            sendBridgeError("Activity not available")
            return
        }
        currentSessionId = ""
        val apiToken = args["apiToken"] as? String ?: ""
        val apiURL = args["apiURL"] as? String ?: ""
        val language = args["language"] as? String ?: "EN"
        val isExportFace = args["isExportFace"] as? Boolean ?: false
        ifEncryption = args["isEncrypt"] as? Boolean ?: false
        val faceImageBase64 = args["faceImageBase64"] as? String
        val isActiveLiveness = args["isActiveLiveness"] as? Boolean ?: false

        if (faceImageBase64 == null) {
            sendBridgeError("Face Verify: faceImageBase64 is required")
            return
        }

        try {
            val faceImageBytes = Base64.decode(faceImageBase64, Base64.DEFAULT)
            android.graphics.BitmapFactory.decodeByteArray(faceImageBytes, 0, faceImageBytes.size)
                ?: throw Exception("Failed to decode image bytes to Bitmap")

            wiseAiApp.init(currentActivity, apiToken, apiURL)
            WiseAiApp.storeFaceImage(faceImageBytes)

            val intent = Intent(currentActivity, WiseAiFaceVerify::class.java).apply {
                putExtra("EXPORT_FACE", isExportFace)
                putExtra("CAMERA_FACING", "FRONT")
                putExtra("IF_ENCRYPTION", ifEncryption)
                putExtra("LANGUAGE_CODE", language)
                putExtra("ACTIVATE_ACTIVE_LIVENESS", isActiveLiveness)
            }

            WiseAiApp.startNewSession(ifEncryption, object : SessionCallback {
                override fun onComplete(data: Any?) {
                    handleSessionCallback(data)
                    currentActivity.startActivityForResult(intent, ACTIVITY_FACE_VERIFY)
                }
                override fun onError(exception: String) {
                    sendBridgeError(exception)
                }
            })
        } catch (e: Exception) {
            sendBridgeError("Face Verify init failed: ${e.message}")
        }
    }

    // MARK: - Activity Result Handling

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        when (requestCode) {
            ACTIVITY_EKYC,
            ACTIVITY_PASSPORT_NFC_EKYC,
            ACTIVITY_FACE_VERIFY -> {
                handleSdkActivityResult(resultCode, data)
                return true
            }
            else -> return false
        }
    }

    private fun handleSdkActivityResult(resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            // Step 1: get the SDK's raw result string. If getResult() throws,
            // there is no SDK response to forward — emit a bridge error.
            val raw: String
            try {
                raw = WiseAiApp.getResult()
            } catch (e: Exception) {
                sendBridgeError("getResult() threw: ${e.message}")
                return
            }

            // Step 2: decrypt if needed. On any decryption failure, fall back
            // to the original raw string so the backend still sees something
            // with sessionId. Do NOT replace it with a custom envelope.
            var payload: String
            try {
                payload = decryptResultIfNeeded(raw)
            } catch (e: Exception) {
                Log.w(TAG, "Decryption failed, forwarding original: ${e.message}")
                payload = raw
            }

            sendSdkResult(payload)
        } else {
            // Non-OK result from the SDK activity: cancellation or internal abort.
            val status = data?.getStringExtra("STATUS")
            if (status == "CANCELLED") {
                sendCancelled()
            } else {
                // SDK activity exited without a structured response — nothing to forward.
                sendBridgeError("SDK activity ended with non-OK result, status=$status")
            }
        }
    }

    // MARK: - Session Callback Helpers
    private fun handleSessionCallback(data: Any?) {
        if (data == null) return
        try {
            val obj = JsonParser.parseString(data.toString()).asJsonObject
            // SDK v2.7.7+: sessionId returned immediately upon session creation
            // regardless of encryption.
            if (obj.has("sessionId")) {
                currentSessionId = obj.get("sessionId").asString ?: ""
            }
            // Encryption keys are only present when isEncrypt = true.
            if (obj.has("key")) {
                key = obj["key"].asString
                iv = obj["iv"].asString
                padding = obj["padding"].asString
                mode = obj["mode"].asString
                encryptionAlgorithm = obj["alg"].asString
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse session callback data: ${e.message}")
        }
    }

    private fun decryptResultIfNeeded(result: String): String {
        var jsonObject = JsonParser.parseString(result).asJsonObject
        if (jsonObject.has("encryptedResult")) {
            val sessionId = jsonObject["sessionId"].asString
            val encryptedText = jsonObject["encryptedResult"].asString
            val decryptedString = decryptString(encryptedText, key, iv, padding, mode, encryptionAlgorithm)
            val decryptedJson = JsonParser.parseString(decryptedString).asJsonObject
            decryptedJson.addProperty("sessionId", sessionId)
            jsonObject = decryptedJson
        }
        return jsonObject.toString()
    }

    @Throws(Exception::class)
    private fun decryptString(
        encryptedText: String,
        key: String,
        initVector: String,
        padding: String,
        mode: String,
        algorithm: String
    ): String {
        val ivSpec = IvParameterSpec(Base64.decode(initVector, Base64.DEFAULT))
        val updatedAlgorithm = if (algorithm == "AES256") "AES" else algorithm
        val sKeySpec = SecretKeySpec(Base64.decode(key, Base64.DEFAULT), updatedAlgorithm)
        val cipher = Cipher.getInstance("$updatedAlgorithm/$mode/$padding")
        cipher.init(Cipher.DECRYPT_MODE, sKeySpec, ivSpec)
        return String(cipher.doFinal(Base64.decode(encryptedText, Base64.DEFAULT)), Charsets.UTF_8)
    }

    // MARK: - Event Senders (the ONLY three exits to Flutter)

    private fun sendSdkResult(rawData: String) = sendEvent("sdk", rawData)
    private fun sendCancelled() = sendEvent("cancelled", "")
    private fun sendBridgeError(message: String) {
        Log.e(TAG, "bridgeError: $message")
        sendEvent("bridgeError", message)
    }

    private fun sendEvent(source: String, rawData: String) {
        eventSink?.let { sink ->
            val event = mutableMapOf<String, Any>(
                "source" to source,
                "rawData" to rawData
            )
            // Attach early sessionId (SDK v2.7.7+) so cancelled / bridgeError
            // events can still be correlated to the session. For "sdk" events,
            // sessionId is also inside rawData; the duplicate is harmless.
            if (currentSessionId.isNotEmpty()) {
                event["sessionId"] = currentSessionId
            }
            sink.success(event)
        } ?: Log.w(TAG, "EventSink is null — dropping event source=$source")
    }
}
